jvb/src/main/kotlin/org/jitsi/videobridge/xmpp/ClientConnection.kt

```diff
@@ -0,0 +1,199 @@
+import org.jitsi.xmpp.util.IQUtils
+import org.jivesoftware.smack.packet.ExtensionElement
+import org.jivesoftware.smack.packet.IQ
+import org.jivesoftware.smack.packet.XMPPError
+import org.jivesoftware.smackx.iqversion.packet.Version
+import org.json.simple.JSONObject
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * The XMPP client connection for the videobridge
+ */
+class ClientConnection : IQListener {
+    private val logger = createLogger()
+
+    /**
+     * The [MucClientManager] which manages the XMPP user connections and the MUCs.
+     */
+    val mucClientManager = MucClientManager(FEATURES)
+
+    private val config = XmppClientConnectionConfig()
+
+    private val running = AtomicBoolean(false)
+
+    var eventHandler: EventHandler? = null
+
+    fun start() {
+        if (running.compareAndSet(false, true)) {
+            mucClientManager.apply {
+                registerIQ(HealthCheckIQ())
+                registerIQ(ColibriConferenceIQ())
+                registerIQ(Version())
+                registerIQ(ShutdownIQ.createForceShutdownIQ())
+                registerIQ(ShutdownIQ.createGracefulShutdownIQ())
+                setIQListener(this@ClientConnection)
+            }
+
+            config.clientConfigs.forEach { cfg -> mucClientManager.addMucClient(cfg) }
+        } else {
+            logger.info("Already started")
+        }
+    }
+
+    fun stop() {
+        if (running.compareAndSet(true, false)) {
+            mucClientManager.stop()
+        }
+    }
+
+    fun setPresenceExtension(extension: ExtensionElement) {
+        mucClientManager.setPresenceExtension(extension)
+    }
+
+    fun addMucClient(jsonObject: JSONObject): Boolean {
+        if (jsonObject["id"] !is String) {
+            return false
+        }
+        val config = MucClientConfiguration(jsonObject["id"] as String)
+        for (key in jsonObject.keys) {
+            val value = jsonObject[key]
+            if (key is String && value is String && key != "id") {
+                config.setProperty(key, value)
+            }
+        }
+        if (!config.isComplete) {
+            logger.info("Not adding a MucClient, configuration incomplete.")
+            return false
+        } else {
+            mucClientManager.addMucClient(config)
+        }
+
+        // We consider the case where a client with the given ID already
+        // exists as success. Note however, that the existing client's
+        // configuration was NOT modified.
+        return true
+    }
+
+    fun removeMucClient(jsonObject: JSONObject): Boolean {
+        if (jsonObject["id"] !is String) {
+            return false
+        }
+        return mucClientManager.removeMucClient(jsonObject["id"] as String)
+    }
+
+    /**
+     * Process an incoming IQ
+     */
+    override fun handleIq(iq: IQ?): IQ? {
+        if (iq == null) {
+            return null
+        }
+        logger.cdebug { "RECV: ${iq.toXML()}" }
+
+        val response = when (iq.type) {
```

Comment:
IQ types are `get`, `set`, `result` and `error`. The latter two are responses, and we wouldn't send anything back for them, so there's not need to log "SENT null".

The code could be simplified some, too: 
```
return when (iq.type) {
  IQ.Type.get, IQ.Type.set -> handleIqRequest(iq).also { log() }
  else -> null
}